docs/yul.rst

```diff
@@ -83,10 +101,301 @@ and ``add`` to be available.
+i.e. in this example, ``y`` is on top of the stack and ``x``
+is below it.
+
+Variable Declarations
+---------------------
+
+You can use the ``let`` keyword to declare variables.
+A variable is only visible inside the
+``{...}``-block it was defined in. When compiling to the EVM,
+a new stack slot is created that is reserved
+for the variable and automatically removed again when the end of the block
+is reached. You can provide an initial value for the variable.
+If you do not provide a value, it will be zero.
+
+Since variables are stored on the stack, they do not directly
+influence memory or storage, but they can be used as pointers
+to memory or storage locations in the built-in functions
+``mstore``, ``mload``, ``sstore`` and ``sload``.
+Future dialects migh introduce specific types for such pointers.
+
+When a variable is referenced, its current value is copied.
+For the EVM, this translates to a ``DUP`` instruction.
+
+.. code::
+
+    {
+        let zero := 0
+        let v := calldataload(zero)
+        {
+            let y := add(sload(v), 1)
+            v := y
+        } // y is "deallocated" here
+        sstore(v, zero)
+    } // v and zero are "deallocated" here
+
+Depending on the settings, the compiler can already
+remove the stack slot after the variable has been used for
+the last time, even though it is still in scope.
+
+Assignments
+-----------
+
+Variables can be re-assigned arbitrarily using the
+``:=`` operator. It is possible to assign multiple
+variables at the same time and the number and types
+of values have to match.
+If you want to assign the values returned from a function that has
+multiple return parameters, you have to provide multiple variables.
+
+.. code::
+
+    let v := 0
+    // re-assign v
+    v := 2
+    let t := add(v, 2)
+    function f() -> a, b { }
+    // assign multiple values
+    v, t := f()
+
+
+If
+--
+
+The if statement can be used for conditionally executing code.
+There is no "else" part, consider using "switch" (see below) if
+you need multiple alternatives.
+
+.. code::
+
+    if eq(value, 0) { revert(0, 0) }
+
+The curly braces for the body are required.
+
+Switch
+------
+
+You can use a switch statement as an extended version of "if/else".
+It takes the value of an expression and compares it to several literal constants.
+The branch corresponding to the matching constant is taken. Contrary to the
+error-prone behaviour of some programming languages, control flow does
+not continue from one case to the next. There can be a fallback or default
+case called ``default`` which is taken if none of the literal constants matches.
+
+.. code::
+
+    {
+        let x := 0
+        switch calldataload(4)
+        case 0 {
+            x := calldataload(0x24)
+        }
+        default {
+            x := calldataload(0x44)
+        }
+        sstore(0, div(x, 2))
+    }
+
+The list of cases does not require curly braces, but the body of a
```

Comment:
```suggestion
The list of cases is not enclosed by curly braces, but the body of a
```
Maybe fine to say "not require", but it's not like you can optionally have them - in fact, I do that wrong *every* time :-D.

Reply:
Actually we might want to change this.

Reply:
https://github.com/ethereum/solidity/issues/8142